onnxruntime/contrib_ops/rocm/fused_conv.cc

```diff
@@ -0,0 +1,321 @@
+                                                     Base::s_.y_data,
+                                                     &beta,
+                                                     Base::s_.y_tensor,
+                                                     Base::s_.y_data));
+    }
+    if (Base::s_.post_slicing_required) {
+      ORT_RETURN_IF_ERROR(onnxruntime::rocm::SliceOutUnwantedOutputSection(
+          this->Stream(),
+          Base::s_.y_data,
+          Base::s_.y_dims_with_adjusted_pads,
+          Base::s_.Y->MutableDataRaw(),
+          Base::s_.y_dims.GetDims(),
+          Base::s_.slice_starts,
+          Base::s_.slice_ends,
+          Base::s_.slice_axes,
+          Base::s_.element_size));
+    }
+    return Status::OK();
+  }
+
+ private:
+  Status MapMode(const std::string& activaton_mode) {
+    if (activaton_mode == "Relu") {
+      activation_mode_ = miopenActivationMode_t::miopenActivationRELU;
+    } else {
+      return Status(common::StatusCategory::ONNXRUNTIME,
+                    common::StatusCode::INVALID_ARGUMENT,
+                    "unsupported conv activation mode");
+    }
+    return Status::OK();
+  }
+  miopenStatus_t status_ = miopenStatusNotInitialized;
+  miopenActivationMode_t activation_mode_;
+  miopenActivationDescriptor_t activation_desc_ = nullptr;
+
+  // MIOpen Fusion API
+  // TODO: create one fusion descriptor shared by multiple FusedConv
+  //       objects
+  //
+  // Considerations:
+  // How to determine two FusedConv objects may share the same fusion
+  // descriptor? Hashing x_tensor,conv_desc, etc.?
+  struct FusedConvFusionData {
+    miopenFusionPlanDescriptor_t plan = nullptr;
+    miopenFusionOpDescriptor_t conv_op = nullptr;
+    miopenFusionOpDescriptor_t bias_b_op = nullptr;
+    miopenFusionOpDescriptor_t bias_z_op = nullptr;
+    miopenFusionOpDescriptor_t act_op = nullptr;
+    miopenOperatorArgs_t fusion_args = nullptr;
+
+    // TODO: There is a potential problem. miopenHandle_t may be destroyed and
+    //       re-created later, sharing the same address. Currently there is any way
+    //       to detect it?
+    mutable std::unordered_set<miopenHandle_t> compiled_on;
+
+    FusedConvFusionData(const FusedConvFusionData&) = delete;
+    FusedConvFusionData& operator= (const FusedConvFusionData&) = delete;
+
+    FusedConvFusionData(FusedConvFusionData&& other) {
+      *this = std::move(other);
+    }
+    FusedConvFusionData& operator=(FusedConvFusionData&& other) {
+      std::swap(this->plan, other.plan);
+      std::swap(this->fusion_args, other.fusion_args);
+      this->conv_op = other.conv_op;
+      this->bias_b_op = other.bias_b_op;
+      this->bias_z_op = other.bias_z_op;
+      this->act_op = other.act_op;
+      this->compiled_on = std::move(other.compiled_on);
+      return *this;
+    }
+
+    FusedConvFusionData() { }
+    ~FusedConvFusionData() {
+      if (plan) {
+        miopenDestroyFusionPlan(plan);
```

Comment:
Not quite familiar with miopenFusionPlan, just want the make sure by destroying this plan and args, those conv_op, bias_b_op, bias_z_op, act_op (if not nullptr) are also destroyed so there is no memory leak, right?

Reply:
Yes, the [document](https://docs.amd.com/bundle/MIOpen_gh-pages/page/Getting_Started_FusionAPI.html#cleanup) stated so
> Once the fusion plan object is destroyed, all the operations created are destroyed automatically and do not need any special cleanup.